snap: make `snap run` look at the system-key for security profiles

mvo5 opened the pull request:
many: make `snap run` look at the system-key for security profiles

If the system-key for the security profiles does not match, wait and
eventually timeout if snapd is not generating the profiles we need.

Without it we may have the following scenario:

1. snapd gets refreshed and snaps need updated security profiles
   to work (e.g. because snap-confine needs new generated
   snap-update-ns profiles on disk)
2. The system reboots to start the new snapd. At this point
   the old security profiles are on disk (because the new
   snapd did not run yet) and nothing new is there yet.
3. Snaps that run as daemon get started during boot by systemd
   (e.g. network-manager). This may happen before snapd had a
   chance to refresh the security profiles.
4. Because the security profiles are for the old version of
   the snaps that run before snapd starts and generates the
   missing security profiles will fail to start. For e.g.
   network-manager this is of course catastrophic as it
   leaves the machine without networking.

To prevent this, in step(4) we have this wait-for-snapd
step to ensure the expected profiles are on disk.

This PR also switches the system-key to json to make the
parsing a lot quicker.

It also introduces a new version key into the system-key.
This version key is important because there is a race when
a) the sytem-key gets new inputs
b) old `snap run` is executed and slow to load
c) new snapd with new inputs run at exactly the same time and
   write a new system-key to disk
In this case (b) will have calculated a profile without the new
inputs and eventually timeout. The new code will just continue
if there is a version mismatch (which is fine because it means
new profiles are on disk so no need to wait).

zyga commented on cmd/snap/cmd_run.go:
```diff
@@ -91,6 +93,40 @@ and environment.
+		return false, err
+	}
+	systemKey := string(raw)
+	return systemKey != interfaces.SystemKey(), nil
```
This means that snap run must be able to compute the system key which is non-trivial in complexity. 

How about this pattern:

- we look at /proc/uptime
- we stat the system key file
- if the system key is more recent than time since boot time then we are good and there's nothing to do
- if the system key is older then assume it is out of date and wait

On snapd side:
- if the system key is correct then open it and update mtime
- in all other cases behave as before.

This way snap run can easily and quickly check that we are in a good state without involving RPC or really high complexity operations.

mvo5 replied:
I think we also have a race with classic when it restarts so I think it is not sufficient to look at uptime. Also (just for the record): the system-key is not trivial but its also not expensive, it does not involve RPC or highly complex operations, it does: look at the build-id (open /proc/self/exe and reading the elf header), reading the apparmor sysfs dir, checking for nfs (reading the mountinfo file), checking if root is on an overlay (reading the mountinfo file), checking core revision (one readlink call), reading the seccomp actions (one read in proc which will happen on startup anyway). So I agree there is overhead but I would not call it highly complex :)

niemeyer commented on cmd/snap/cmd_run.go:
```diff
@@ -91,6 +93,40 @@ and environment.
+		time.Sleep(1 * time.Second)
+	}
+
+	return fmt.Errorf("cannot find updated security profiles after system-key change")
```
"timeout waiting for snap system profiles to get updated"

mvo5 commented:
Thanks for the review and the HO yesterday. I updated the PR (squashed/rebased for easier merging into 2.32, sorry for this but I think because the amount of writen feedback in here was small that was not too terrible).

I addressed the points we discussed:
- moved to json
- add version attribute to the system-key
- make wait 200ms per iteration
- deal better with errors (i.e. try to run it anyway on error)

mvo5 commented:
Setting to blocked because of course `snap run` must read the build-id of the current snapd. So it needs to figure out which snapd to look at (re-exec etc) and use that build-id as input for the system-key.

pedronis commented on cmd/snap/cmd_run.go:
```diff
@@ -91,6 +92,33 @@ and environment.
+	//
+	// (wait up to 1500 * 200ms = 300s)
+	timeout := 1500
+	if timeoutEnv := os.Getenv("SNAPD_DEBUG_SYSTEM_KEY_WAIT_TIMEOUT"); timeoutEnv != "" {
```
I understand is only for tests but is a bit strange that this envvar is not in s or ms but  in # 200ms

pedronis commented on interfaces/system_key.go:
```diff
@@ -20,28 +20,51 @@
+	BuildIDs struct {
+		HostSnapd     string `json:"host_snapd"`
+		CoreSnapSnapd string `json:"core_snap_snapd"`
+	} `json:"build_ids"`
```
was it discussed to use both like this?  it will provoke some more regens than strictly necessary

mvo5 replied:
I will added the following comment:
       // The build-ids of both the core snap snapd and the host
	// snapd are used as inputs. This may lead to more regens
	// than strictly necessary. The alternative is to try to
	// figure out which one will be run, this means to look
	// for re-exec behaviour and for potential SNAP_REEXEC=0
	// in the snapd service file or /etc/environment. So
	// this approach is conservative by looking at both
	// at the expense of potential extra re-gens.

pedronis replied:
to be clear it means  that switching a system that does reexec, to SNAP_REEXEC=0 will not regen profiles (unless because of Version?), I suppose we don't really support turning off reexec except in a very controlled way (the fact that is just a envvar from that POV is a bit misleading)

pedronis commented on interfaces/system_key.go:
```diff
@@ -20,28 +20,51 @@
-	Core             string   `yaml:"core,omitempty"`
-	SecCompActions   []string `yaml:"seccomp-features"`
+	// IMPORTANT: when adding new inputs bump this version
+	Version string `json:"version"`
```
why not just a int?

pedronis commented on interfaces/system_key.go:
```diff
@@ -93,26 +124,83 @@ func generateSystemKey() *systemKey {
+		if exe, err := os.Readlink("/proc/self/exe"); err == nil {
+			// detect running local local builds
+			if !strings.HasPrefix(exe, "/usr") && !strings.HasPrefix(exe, "/snap") {
+				return false, ErrSystemKeyIncomparableVersions
```
we should probably log something here

pedronis commented on interfaces/system_key.go:
```diff
@@ -93,26 +124,83 @@ func generateSystemKey() *systemKey {
+		}
+	}
+
+	return string(mySystemKeyJSON) != string(raw), nil
```
didn't Gustavo suggest to use  reflect.DeepEqual instead of comparing the serialized JSON?

mvo5 replied:
Happy to do this.

pedronis commented on interfaces/system_key.go:
```diff
@@ -93,26 +124,83 @@ func generateSystemKey() *systemKey {
+	// inputs than the "snap run" in memory. In this case we
+	// should be fine because new security profiles will also
+	// have been written to disk.
+	if mySystemKey.Version != diskSystemKey.Version {
```
do we need to differentiate  increasing vs decreasing versions case?

mvo5 replied:
I don't think so.

niemeyer replied:
Also don't think so.. if it's different we can't compare.

pedronis commented on overlord/ifacestate/helpers.go:
```diff
@@ -114,22 +110,12 @@ func (m *InterfaceManager) addSnaps() error {
 }
 
 func (m *InterfaceManager) profilesNeedRegeneration() bool {
```
this is run too early if we are seeded,  and it's a problem even on classic, we probably need to return false if we are not seeded, and trigger a variant of writing the key when we are about to setup profiles for core for the first time?

pedronis commented on cmd/snap/cmd_run.go:
```diff
@@ -91,6 +92,42 @@ and environment.
+	// connect we just continue because that means snapd has
+	// generated new profiles.
+	//
+	// connect timeout is 5s
```
are we sure 5s is enough on something slow?

mvo5 replied:
I clarified the comment, my reading of the client.doSync is that it waits for up to 5s to connect to snapd. I hope the updated comment makes this clear.

pedronis commented on cmd/snap/cmd_run.go:
```diff
@@ -91,6 +92,42 @@ and environment.
+	for i := 0; i < timeout; i++ {
+		if _, err := cli.SysInfo(); err == nil {
+			return nil
+		}
```
shouldn't we sleep a bit here?

mvo5 replied:
Yes, I think this makes sense, cli.SysInfo() should sleep when it cannot connect but we should also do it for good measure.

pedronis commented on interfaces/system_key.go:
```diff
@@ -20,47 +20,86 @@
 	mockedSystemKey *systemKey
 )
 
+func findSnapdPath() (string, error) {
```
not for now but I think we have various functions like this, maybe we can find a place for them? dirs?

mvo5 replied:
Good point. I will see if I can squeeze it in here or into a followup.

zyga commented on interfaces/system_key.go:
```diff
@@ -20,47 +20,86 @@
 
+// ErrSystemKeyIncomparableVersions indicates that the system-key
+// on disk and the system-key calculated from generateSystemKey
+// have different inputs and are therefor incomparable.
```
I think you want to say `therefore`

zyga commented on interfaces/system_key.go:
```diff
@@ -20,47 +20,86 @@
 // running system is similar enough to the generated profiles or
 // if the profiles need to be re-generated to match the new system.
+//
+// Note that this key gets generated on *each* `snap run` - so it
```
Please highlight that the key is generated multiple times in case we wait.

mvo5 replied:
We no longer generate it multiple times, on the first mismatch we now try to reach snapd.

zyga replied:
Ah, thank you for correcting that!

zyga commented on interfaces/system_key.go:
```diff
@@ -20,47 +20,86 @@
+	// IMPORTANT: when adding new inputs bump this version
+	Version int `json:"version"`
+
+	BuildID          string   `json:"build-id"`
```
Can you please add a newline below BuildID and add this comment
```
// The elements below are external to snapd and change depending on
// the running kernel and system configuration.
```

niemeyer replied:
Having a comment here is good, but that comment is saying the obvious: that these settings change when the system changes.

I think what we need to say is that these settings are here because they may cause profiles to be built in a different way, which may affect the ability of snap applications to run correctly after snapd (and thus snap run/snap-confine/etc) are updated, but before the profiles are generated.

mvo5 replied:
I updated the comment now, hope it adds clarity.

pedronis commented on interfaces/system_key.go:
```diff
@@ -80,39 +119,86 @@ func generateSystemKey() *systemKey {
+	}
+	var diskSystemKey systemKey
+	if err := json.Unmarshal(raw, &diskSystemKey); err != nil {
+		return false, err
```
because previous versions of snapd were writing yaml, will we not hit this which means we will fail to wait for the new profiles on 2.31->2.32?

mvo5 replied:
Yes, this is rather critical indeed. I think because we now have a second line of defense by talking to snapd we can simply on error try to reach snapd. This should be ok in the yaml -> json transition as it means that by the time we reach the new snapd the transition will have happend or at least we don't block because we can reach snapd.

zyga commented on interfaces/system_key.go:
```diff
@@ -20,47 +20,86 @@
+		Version: 1,
+	}
+
+	if snapdPath, err := findSnapdPath(); err == nil {
```
Should we log either of the errors in case they happen?

mvo5 replied:
This is now returning an error.

zyga commented on interfaces/system_key.go:
```diff
@@ -80,39 +119,86 @@ func generateSystemKey() *systemKey {
+func WriteSystemKey() error {
+	sks, err := json.Marshal(generateSystemKey())
+	if err != nil {
+		panic(err)
```
Why not return the error if we have the chance?

niemeyer replied:
+1, let's please not panic.

mvo5 replied:
Yes, good point!

zyga commented on interfaces/system_key.go:
```diff
@@ -80,39 +119,86 @@ func generateSystemKey() *systemKey {
+// To prevent this, in step(4) we have this wait-for-snapd
+// step to ensure the expected profiles are on disk.
+func SystemKeyMismatch() (bool, error) {
+	mySystemKey := generateSystemKey()
```
AFAIR we said that system key generation should be able to fail.

niemeyer replied:
Indeed, and this function (SystemKeyMismatch) already presents that interface. We just need to make generateSystemKey able to report the failure too.

mvo5 replied:
Indeed.

zyga commented on overlord/ifacestate/helpers.go:
```diff
@@ -114,22 +110,12 @@ func (m *InterfaceManager) addSnaps() error {
 	if err != nil {
-		logger.Noticef("cannot read system-key file: %s", err)
+		logger.Noticef("error trying to compare the snap system key: %v", err)
 		return true
```
This is fine. The re-generation is doing as little work as needed anyway.

niemeyer commented on interfaces/system_key.go:
```diff
@@ -20,47 +20,86 @@
-	buildID, err := osutil.MyBuildID()
-	if err != nil {
-		buildID = ""
+	var err error
```
This variable is strange here. It's not used by any of the code introduced in the changes in this PR.

niemeyer commented on interfaces/system_key.go:
```diff
@@ -80,39 +119,86 @@ func generateSystemKey() *systemKey {
+		}
+	}
+
+	return !reflect.DeepEqual(mySystemKey, &diskSystemKey), nil
```
Since we have actual structs, we might compare them with code, but that's good enough as a starting point.

zyga commented on interfaces/system_key.go:
```diff
@@ -20,45 +20,92 @@
+	// These inputs come from the host environment via e.g.
+	// kernel version or similar settings. If those change we may
+	// need to change the generated profiles (e.g. when the user
+	// boots into a more featureful seccomp).
```
I would just say "featureful kernel" instead of "featureful seccomp"

mvo5 replied:
featureful kernel!
